test/kt/single_pass_copy_if.cpp

```diff
@@ -61,13 +61,18 @@ generate_copy_if_data(T* input, std::size_t size, std::uint32_t seed)
 {
     // Integer numbers are generated even for floating point types in order to avoid rounding errors,
     // and simplify the final check
-    using substitute_t = std::conditional_t<std::is_signed_v<T>, std::int64_t, std::uint64_t>;
-
     std::default_random_engine gen{seed};
-    substitute_t start = std::is_signed_v<T> ? -10 : 0;
-    substitute_t end = std::is_signed_v<T> ? 10 : 20;
-    std::uniform_int_distribution<substitute_t> dist(start, end);
-    std::generate(input, input + size, [&] { return dist(gen); });
+
+    if constexpr (std::is_integral_v<T>)
+    {
+        std::uniform_int_distribution<T> dist(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+        std::generate(input, input + size, [&] { return dist(gen); });
+    }
+    else
+    {
+        std::uniform_real_distribution<T> dist(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+        std::generate(input, input + size, [&] { return dist(gen); });
+    }
```

Comment:
Two things to improve:

- Use `lowest` instead of `min` (`min` is ~0 for float types).
- Make sure that duplicates are generated.

You can take a look at `test\kt\esimd_radix_sort_utils.h:generate_data()`, it is done there.

Reply:
Yes, good catch with `lowest()`, I _know_ that but don't always _remember_ that :) .

I don't understand why it is interesting to have duplicates for `copy_if` with this "less than cutoff" predicate, for radix sort it makes sense, but I don't see the need here.

I dont mind just including this utility header and call this generation function, no need to duplicate this functionality.  While I don't understand the need for duplicates, they don't harm anything here.

Since we may use it here, can you explain the motivation for generating real numbers as it is done in `test\kt\esimd_radix_sort_utils.h:generate_data()`? Specifically:
1. Why limit to `1e12` rather than `numerical_limits<T>::max()`?
2. Why separate the sign into a separate distribution rather than `std::uniform_real_distribution<T> dist_real(-log2(1e12), log2(1e12));`?
3. Why use `log2` followed by `exp2` ?

If they perhaps have limited interest to things outside of radix sort, we may be able to make them tunable parameters.

Reply:
> I don't understand why it is interesting to have duplicates for copy_if with this "less than cutoff" predicate, for radix sort it makes sense, but I don't see the need here.

Yeah, you are right, it's not relevant for `copy_if`. Sorry to bother you with that.

> Why limit to 1e12 rather than numerical_limits<T>::max()?

The distribution is uniform and floating point types can hold very large values, so the values generated with `max` will gravitate towards edges. This is what I got: `9.43047e+307 4.73343e+307 3.52862e+307 9.20991e+307 4.6219e+307 1.466e+308 8.12609e+307 4.44764e+307`.

The trick with exponent allows making a more fair distribution in terms of underlying bit representation (again, it is useful for radix sort), and it appears to provide more real-world values. This is what I got with `log2(1e12)` (almost 40, basically): `-1.97263e+06 32710.5 -1.40546e+06 -19.2773 265671 181.673 -4.88076e+09 6.98599e+08`. 

> Why separate the sign into a separate distribution rather than std::uniform_real_distribution<T> dist_real(-log2(1e12), log2(1e12));

That's because of `exp2` usage below. It provides only positive values. 

> Why use log2 followed by exp2

Imaging that with the exponent transformation, each green tick on the picture below has equal probability:

![image](https://github.com/oneapi-src/oneDPL/assets/69916350/b5a55d0f-4b72-4bc0-811a-0cf2334404c7)
 
This picture is from https://en.wikipedia.org/wiki/Floating-point_arithmetic. It mimics the distribution of representable values of a floating point type.

It would be better to use `log2(numeric_limits::max)` instead of `log2(1e12)` and leave a comment, though. `1e12` is the maximum value after log2 and exp2 transformations. This pair is complementary and that is why it is used together.

Meanwhile, I also checked `std::numeric_limits<T>::lowest(), std::numeric_limits<T>::max()` and got a sequence of `inf` values: https://godbolt.org/z/YzzrzzY6e.

Reply:
OK I think I understand, I suppose for floating point it makes sense to "center" the values on some smaller range, rather than being very likely to end up with gigantic positive or negative values.  I can buy the argument that this is more likely to be similar to a real case.

If I understand correctly, the `log2` `exp2` is to provide a distribution which is more uniform in bit representation for radix sort.  Its actually an interesting question if this is the "correct" thing to do when testing or timing radix sort for floating point numbers.  This will make it act most like a uniformly distributed integral type, but one could also argue that real world data "doesn't care" about bit distribution and that just making a uniform distribution of real numbers is more appropriate, for instance uniform(-1e12,1e12).  On the other hand, I have no real evidence that one distribution is more likely in user code than another for floating point data.  

Very interesting that you get `inf` numbers when using `lowest()` and `max()`.  Seems like a bug, but I can also understand why using the type's limits in the distribution would make for difficult calculations that may involve the "size" of the range.

I will try to unify the data generation, and perhaps make optional some of the radix sort specific stuff.

Reply:
Thanks, Dan. I agree with your suggestions.

I guess it is fine to have such a distribution for tests to cover as many bits combinations as possible, assuming the generated data does not look too artificial. This is because radix sort does many bit-related manipulations, and it should help to catch cases like this one: https://github.com/oneapi-src/oneDPL/blob/b47ff69d059e72384b01df040423c916f21ca835/include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_utils.h#L100-L104

The benchmarks is a different story and `uniform(-1e12,1e12)` looks better. As for copy_if, ` uniform(-1e12,1e12)` looks better too.